File: pkg/browser/playwright/keyboard.go
package playwright

import (
	"testflowkit/pkg/browser"

	pw "github.com/playwright-community/playwright-go"
)

var keyMap = map[browser.Key]string{
	browser.KeyEnter:      "Enter",
	browser.KeyTab:        "Tab",
	browser.KeyDelete:     "Delete",
	browser.KeyEscape:     "Escape",
	browser.KeySpace:      " ",
	browser.KeyArrowUp:    "ArrowUp",
	browser.KeyArrowRight: "ArrowRight",
	browser.KeyArrowDown:  "ArrowDown",
	browser.KeyArrowLeft:  "ArrowLeft",
}

type playwrightKeyboard struct {
	page pw.Page
}

func (k *playwrightKeyboard) Press(key browser.Key) error {
	playwrightKey, ok := keyMap[key]
	if ok {
		return k.page.Keyboard().Press(playwrightKey)
	}

	if len(key) == 1 {
		return k.page.Keyboard().Press(string(key))
	}

	return &UnknownKeyError{Key: key}
}

func newPlaywrightKeyboard(page pw.Page) browser.Keyboard {
	return &playwrightKeyboard{page: page}
}

// UnknownKeyError represents an error when an unknown key is pressed.
type UnknownKeyError struct {
	Key browser.Key
}

func (e *UnknownKeyError) Error() string {
	return "unknown key: " + string(e.Key)
}

File: packages/testflowkit-ts/src/steps/keyboard.ts
import { When } from '@cucumber/cucumber';
import type { TestFlowKitWorld } from '../world.js';

const KEY_MAP: Record<string, string> = {
  Enter: 'Enter',
  Tab: 'Tab',
  Delete: 'Delete',
  Escape: 'Escape',
  Space: 'Space',
  'Arrow Up': 'ArrowUp',
  'Arrow Right': 'ArrowRight',
  'Arrow Down': 'ArrowDown',
  'Arrow Left': 'ArrowLeft',
};

// Registered once under When: Cucumber matches step text against every registered
// pattern regardless of the Given/When/Then keyword used in the .feature file, so this
// also works written as Given/Then/And/But without a second registration.
When('the user presses the {string} key', async function (this: TestFlowKitWorld, key: string) {
  const playwrightKey = KEY_MAP[key];
  if (!playwrightKey) {
    throw new Error(`Unsupported key "${key}". Supported keys: ${Object.keys(KEY_MAP).join(', ')}`);
  }
  await this.page.keyboard.press(playwrightKey);
});
